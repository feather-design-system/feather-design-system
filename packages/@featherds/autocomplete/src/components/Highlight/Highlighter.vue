<template>
  <span>
    <span>{{ beginning }}</span>
    <span class="highlight" v-if="highlighted">{{ highlighted }}</span>
    <span v-if="end">{{ end }}</span>
  </span>
</template>

<script>
import Highlight from "./HighlightMixin";
import Highlighter from "./HighlighterMixin";
export default {
  mixins: [Highlight, Highlighter],
  props: {
    text: {
      type: String,
      required: true,
    },
  },
  computed: {
    index() {
      if (this.query && this.query.length === 0) {
        return -1;
      }
      switch (this.highlight.toLowerCase()) {
        case "ignore-case":
          return this.text.toLowerCase().indexOf(this.query.toLowerCase());
        default:
          return -1;
      }
    },
    beginning() {
      if (this.index === -1) {
        return this.text;
      }
      return this.text.slice(0, this.index);
    },
    highlighted() {
      if (this.index === -1) {
        return;
      }
      return this.text.slice(this.index, this.index + this.query.length);
    },
    end() {
      if (this.index === -1) {
        return;
      }
      return this.text.slice(this.index + this.query.length);
    },
  },
};
</script>
<style lang="scss" scoped>
@import "~@featherds/styles/themes/variables";
.highlight {
  font-weight: var($font-bold);
}
</style>
