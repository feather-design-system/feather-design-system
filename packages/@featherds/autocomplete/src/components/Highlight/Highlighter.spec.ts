import Highlighter from "./Highlighter.vue";
import { shallowMount } from "@vue/test-utils";
import { describe, it, expect } from "vitest";
const getWrapper = function (options = {}) {
  return shallowMount(Highlighter, options);
};

describe("Highlighter", () => {
  describe("ignore case", () => {
    it("should highlight text in the middle of the string", () => {
      const wrapper = getWrapper({
        props: {
          text: "abcd",
          highlight: "ignore-case",
          query: "bc",
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it("should highlight text at the start of the string", () => {
      const wrapper = getWrapper({
        props: {
          text: "abcd",
          highlight: "ignore-case",
          query: "ab",
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it("should highlight text at the end of the string", () => {
      const wrapper = getWrapper({
        props: {
          text: "abcd",
          highlight: "ignore-case",
          query: "cd",
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it("should not highlight when there isn't a match", () => {
      const wrapper = getWrapper({
        props: {
          text: "abcd",
          highlight: "ignore-case",
          query: "xx",
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it("should highlight reglardless of case", () => {
      const wrapper = getWrapper({
        props: {
          text: "aBCd",
          highlight: "ignore-case",
          query: "bc",
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
  it("should not highlight when turned off", () => {
    const wrapper = getWrapper({
      props: {
        text: "abcd",
        highlight: "off",
        query: "bc",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
