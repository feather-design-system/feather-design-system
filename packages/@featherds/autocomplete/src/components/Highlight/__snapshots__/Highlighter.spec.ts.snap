// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlighter ignore case should highlight reglardless of case 1`] = `
<span>
  <span>
    a
  </span>
  <span
    class="highlight"
  >
    BC
  </span>
  <span>
    d
  </span>
</span>
`;

exports[`Highlighter ignore case should highlight text at the end of the string 1`] = `
<span>
  <span>
    ab
  </span>
  <span
    class="highlight"
  >
    cd
  </span>
  <!--v-if-->
</span>
`;

exports[`Highlighter ignore case should highlight text at the start of the string 1`] = `
<span>
  <span />
  <span
    class="highlight"
  >
    ab
  </span>
  <span>
    cd
  </span>
</span>
`;

exports[`Highlighter ignore case should highlight text in the middle of the string 1`] = `
<span>
  <span>
    a
  </span>
  <span
    class="highlight"
  >
    bc
  </span>
  <span>
    d
  </span>
</span>
`;

exports[`Highlighter ignore case should not highlight when there isn't a match 1`] = `
<span>
  <span>
    abcd
  </span>
  <!--v-if-->
  <!--v-if-->
</span>
`;

exports[`Highlighter should not highlight when turned off 1`] = `
<span>
  <span>
    abcd
  </span>
  <!--v-if-->
  <!--v-if-->
</span>
`;
